### DISCUSSION TITLE
Redirection fails with Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client

### DISCUSSION THREAD

**Discussion by fabien-h** [vercel/next.js#14547] [Help]
# Here is my redirection code server side

```
export const getServerSideProps = async (
  ctx: GetServerSidePropsContext,
): Promise<unknown> => {
  const data: Data = { a: 10 };

  if (ctx.query.logged !== 'true') {
    ctx.res.setHeader('Location', '/app/login?redirected=true');
    ctx.res.statusCode = 302;
    ctx.res.end();
    return;
  }

  return {
    props: {
      data,
    },
  };
};
```

For the demo, the logged state is simply `?logged=true` in the url.

## Events

- When I reload the page `/app?logged=true`, I stay on the page.
- When I reload the page `/app`, I'm redirected to the login page
- When I come from another page with a `<Link...`, I get this error.

I'm not sure to understand why. There is a `return` after the `res.end()`;

Is there a way to know if the call to the data is just the initial call or if it comes from the client ?

I tried to redirect client side, but next calls the data before I get a chance to do so.

I'm starting to think that the right way would be to have two components:
- One client side component with a loader that display a loader while the data are fetching and then the view when it's loaded
- On server side component that simple display the view with the data synchronously

## Hint (or is it ?) : TypeError: Cannot convert undefined or null to object

When I reload the page `/app`, I'm redirected to the login page. **BUT** I get this error in the console server-side:

```
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at renderToHTML (--web/node_modules/next/dist/next-server/server/render.js:39:78)
    at async --/web/node_modules/next/dist/next-server/server/next-server.js:68:285
    at async --/web/node_modules/next/dist/next-server/server/next-server.js:67:142
    at async DevServer.renderToHTMLWithComponents (--/web/node_modules/next/dist/next-server/server/next-server.js:91:254)
    at async DevServer.renderToHTML (--/web/node_modules/next/dist/next-server/server/next-server.js:92:254)
    at async DevServer.renderToHTML (--/web/node_modules/next/dist/server/next-dev-server.js:22:539)
    at async DevServer.render (--/web/node_modules/next/dist/next-server/server/next-server.js:48:236)
    at async Object.fn (--/web/node_modules/next/dist/next-server/server/next-server.js:35:852)
    at async Router.execute (--/web/node_modules/next/dist/next-server/server/router.js:28:28)
    at async DevServer.run (--/web/node_modules/next/dist/next-server/server/next-server.js:44:494)
    at async DevServer.handleRequest (--/web/node_modules/next/dist/next-server/server/next-server.js:13:133)
```

**Comment by fabien-h** [comment]
The type error is solved by adding a prop object after the `ctx.res.end()`.

```
  if (!ctx.query.logged) {
    ctx.res.setHeader('Location', '/app/login?redirected=true');
    ctx.res.statusCode = 302;
    ctx.res.end();
    return { props: {} };
  }
```

I'm not sure why.

**Reply by paambaati** [reply]
I noticed the same issue, and I’m currently working around it by doing a client-side reduct, even though it isn’t ideal.

And about the `TypeError` — Next.js will always expect you to return `props` from your `getServerSideProps`, so returning them is a proper fix.

**Reply by paambaati** [reply]
https://github.com/vercel/next.js/discussions/11281#discussion-536 is a much larger discussion of this same issue.

**Comment by karlhorky** [comment]
It seems the RFC by Tim is the way to go for the future: https://github.com/vercel/next.js/discussions/14890

**Comment by karlhorky** [comment]
**Edit:** No need to use the solutions on this page anymore - instead use the built-in feature of Next.js to return an object with a key named `redirect` in `getStaticProps` or `getServerSideProps`: [`redirect` in Data Fetching docs](https://nextjs.org/docs/basic-features/data-fetching#:~:text=export%20async%20function%20getserversideprops(context)%20%7B%20const%20res%20%3D%20await%20fetch(%60https%3A%2F%2F...%2Fdata%60)%20const%20data%20%3D%20await%20res.json()%20if%20(!data)%20%7B%20return%20%7B%20redirect%3A%20%7B%20destination%3A%20'%2F'%2C%20permanent%3A%20false%2C%20%7D%2C%20%7D%20%7D)

Seems like this workaround by @samstr could offer a way until the RFC is ready (checking to see whether the `referer` is there):

### `pages/current-season.js`

```js
import { browserRedirect, serverRedirect } from 'utils/redirects';

const redirect = {
  href: '/seasons/[season]/matches',
  asPath: `/seasons/22/matches`,
  permanent: false,
};

export default browserRedirect(redirect);
export const getServerSideProps = async (ctx) => serverRedirect(ctx, redirect);
```

### `utils/redirects.js`

```js
import Router from 'next/router';
import React, { Component } from 'react';

export const browserRedirect = (redirect) => {
  return class Redirect extends Component {
    componentDidMount() {
      const { href, asPath } = redirect;
      Router.push(href, asPath); // Router.replace might be better here? not sure
    }

    render() {
      return <p>Redirecting...</p>;
    }
  };
};

export const serverRedirect = (ctx, redirect) => {
  const { referer } = ctx.req.headers; // if there's no referer then it was a server request
  const { asPath, permanent, external } = redirect;

  if (!referer || external) {
    res.setHeader('Location', asPath);
    res.statusCode = permanent ? 301 : 307;
  }

   return { props: {} };
};
```

Source: https://github.com/vercel/next.js/discussions/11281#discussioncomment-2384

**Reply by aliakseiherman** [reply]
@fabien-h 
I've managed to get it working by removing `res.end()`. Instead of this line I return empty object from `getInitialProps` — this means you need to try and return something like empty props `{ props: {} }` from `getServerSideProps`.
Please note that we do NOT remove `ctx.res.setHeader` and `ctx.res.statusCode`.

**Reply by ypresto** [reply]
It will cause timeout error on production.  `if (process.env.NODE_ENV !== 'development') ctx.res.end()` to fix it.

**Reply by karlhorky** [reply]
You can just use the official `redirect` solution now:

```js
export async function getServerSideProps(context) {
  const res = await fetch(`https://.../data`)
  const data = await res.json()

  if (!data) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    }
  }

  return {
    props: {}, // will be passed to the page component as props
  }
}
```

Source: https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

**Reply by deschantkn** [reply]
I don't know if I am alone but the official redirect creates an infinite loop for me. It keeps running `getServerSideProps` and redirecting indefinitely.

**Reply by karlhorky** [reply]
I've used this successfully in a couple of projects with no infinite loops, shouldn't be an issue.

In which file is your `getServerSideProps` and where is the redirect `destination` set to? If they are pointing to the same file, then yeah, this would create a loop...

**Reply by deschantkn** [reply]
Yeah it's probably something wrong with my code. The `getServerSideProps` is in a private page as it should be and I am redirecting to my login page when the user is not authenticated. It redirects to the login page successfully but `getServerSideProps` somehow keeps running from the previous page

**Reply by karlhorky** [reply]
Do you also have a `getServerSideProps` with a `redirect` in the login page? If so, maybe that's what's running?

**Reply by deschantkn** [reply]
I was definitely the one doing something wrong! I'm still not sure what I was doing wrong but after a good nigh't sleep the redirect seems to be working fine now. Thanks for your help though!

**Reply by ypresto** [reply]
> It will cause timeout error on production. `if (process.env.NODE_ENV !== 'development') ctx.res.end()` to fix it.

At least in my environment with next v10.2.3 the workaround no longer needed, no ERR_HTTP_HEADERS_SENT error with `res.end()`.
Moreover that workaround causes timeout in cypress (End-to-End) test.

**Reply by tlehwalder** [reply]
Is it possible to pass props or context to the redirected page?

This doesn't work:
```javascript
return {
  redirect: {
    destination: '/info',
    permanent: false,
  },
  props: {
    message: "Some Info Message"
  }
}
```

**Comment by mamadfar** [comment]
I have the same problem when i move to a dynamic route
do someone solve this problem?!

**Comment by dungtran5** [comment]
The problem still exist 😢  .....

**Comment by iambherulal** [comment]
To fix this you can add return before res.status().send()

```javascript
if (total > 0) {
  return res.status(200).send({
    status: 0,
    message: "Messages available",
  });
}
return res.status(503).send({
  status: 1,
  message: "Messages not available!",
});
```

Or you can use if..else in this particular code snippet.

```javascript
if (total > 0) {
  res.status(200).send({
    status: 0,
    message: "Messages available",
  });
} else {
  res.status(503).send({
    status: 1,
    message: "Messages not available!",
  });
}
```

**Comment by bijaydas** [comment]
Is this issue solved? I am also facing the same issue

```javascript
db.all(query, credentials, (err, rows) => {
    if (err) {
      return res.status(400).send({ err });
    }
    if (rows.length === 0) {
      return res.status(400).json({
        message: 'Invalid credentials',
      });
    }
    return res.status(200).json({
      message: 'Login success',
    });
  });
```
